xla/service/elemental_ir_emitter.cc

```diff
@@ -534,6 +545,307 @@ llvm::Value* EmitF8e4m3b11fnuzToF16(llvm::Value* f8_value,
+  // right, and how much to rebias the exponent.
+  llvm::Value* zero_count =
+      b->CreateBinaryIntrinsic(llvm::Intrinsic::ctlz, mantissa,
+                               llvm::ConstantInt::get(b->getInt1Ty(), 0x0u));
+  zero_count =
+      b->CreateSub(zero_count, llvm::ConstantInt::get(b->getInt16Ty(), 5));
+  llvm::Value* shift =
+      b->CreateSub(llvm::ConstantInt::get(b->getInt16Ty(), 8), zero_count);
+
+  // Shift the mantissa to the right so that the msb is at bit 2.
+  mantissa = b->CreateLShr(mantissa, shift);
+  exponent = b->CreateSub(exponent, zero_count);
+
+  // Correct for output denorm.
+  llvm::Value* output_norm =
+      b->CreateICmpSGE(exponent, llvm::ConstantInt::get(b->getInt16Ty(), -15));
+  shift = b->CreateSub(llvm::ConstantInt::get(b->getInt16Ty(), -15), exponent);
+  llvm::Value* mantissa_denorm = b->CreateLShr(mantissa, shift);
+
+  mantissa = b->CreateSelect(output_norm, mantissa, mantissa_denorm);
+  exponent = b->CreateSelect(output_norm, exponent,
+                             llvm::ConstantInt::get(b->getInt16Ty(), -16));
+
+  // Shift the sign down from bit 15 to bit 7.
+  llvm::Value* result = b->CreateLShr(sign, 8);
+  // Bias the exponent for the output.
+  exponent =
+      b->CreateAdd(exponent, llvm::ConstantInt::get(b->getInt16Ty(), 16));
+  // Shift it into position.
+  exponent = b->CreateShl(exponent, 2);
+  result = b->CreateOr(result, exponent);
+
+  mantissa = b->CreateAnd(mantissa, 0x3u);
+  result = b->CreateOr(result, mantissa);
+  // Truncate down to int8.
+  result = b->CreateTrunc(result, b->getInt8Ty());
+
+  // Select based on whether the value was in range.
+  return b->CreateSelect(out_of_range_pred,
+                         llvm::ConstantInt::get(b->getInt8Ty(), 0x80u), result);
+}
+
+llvm::Value* EmitF8e5m2fnuzToF16(llvm::Value* f8_value, llvm::IRBuilder<>* b,
+                                 llvm::Module* module) {
+  llvm::ArrayType* result_lut_array_type =
+      llvm::ArrayType::get(b->getInt16Ty(), 128);
+
+  llvm::Constant* global_result_lut_array = module->getOrInsertGlobal(
+      "f8e5m2fnuzToF16LUT", result_lut_array_type,
+      [&]() -> llvm::GlobalVariable* {
+        // Since the function range is only 2^8 and symmetric on the sign bit,
+        // this is implemented as a table lookup.
+        llvm::Constant* result_lut[128];
+
+        // Populate the table with values computed using llvm APFloat.
+        for (uint8_t i = 0; i < 128; ++i) {
+          llvm::APFloat value(llvm::APFloat::Float8E5M2FNUZ(),
+                              llvm::APInt(8, i));
+
+          bool losesInfo;
+          value.convert(llvm::APFloat::IEEEhalf(),
+                        llvm::APFloat::rmNearestTiesToEven, &losesInfo);
+
+          result_lut[i] = llvm::ConstantInt::get(
+              b->getInt16Ty(),
+              static_cast<uint16_t>(value.bitcastToAPInt().getRawData()[0] &
+                                    0x7FFF));
```

Comment:
Calling getZExtValue is more elegant than getRawData. Also, no reason to remove the sign bit since it's already guaranteed not to be set.